core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -142,72 +176,140 @@ public TermsQueryBuilder disableCoord(boolean disableCoord) {
+     * Sets the type name to lookup the terms from.
      */
     public TermsQueryBuilder lookupType(String lookupType) {
-        this.lookupType = lookupType;
+        if (lookupType == null) {
+            throw new IllegalArgumentException("Lookup type cannot be set to null");
+        }
+        if (this.termsLookup == null) {
+            this.termsLookup = new TermsLookup();
+        }
+        this.termsLookup.type(lookupType);
         return this;
     }
 
     /**
-     * Sets the doc id to lookup the terms from.
+     * Sets the document id to lookup the terms from.
      */
     public TermsQueryBuilder lookupId(String lookupId) {
-        this.lookupId = lookupId;
+        if (lookupId == null) {
+            throw new IllegalArgumentException("Lookup id cannot be set to null");
+        }
+        if (this.termsLookup == null) {
+            this.termsLookup = new TermsLookup();
+        }
+        this.termsLookup.id(lookupId);
         return this;
     }
 
     /**
-     * Sets the path within the document to lookup the terms from.
+     * Sets the path name to lookup the terms from.
      */
     public TermsQueryBuilder lookupPath(String lookupPath) {
-        this.lookupPath = lookupPath;
+        if (lookupPath == null) {
+            throw new IllegalArgumentException("Lookup path cannot be set to null");
+        }
+        if (this.termsLookup == null) {
+            this.termsLookup = new TermsLookup();
+        }
+        this.termsLookup.path(lookupPath);
         return this;
     }
 
+    /**
+     * Sets the routing to lookup the terms from.
+     */
     public TermsQueryBuilder lookupRouting(String lookupRouting) {
-        this.lookupRouting = lookupRouting;
+        if (lookupRouting == null) {
+            throw new IllegalArgumentException("Lookup routing cannot be set to null");
+        }
+        if (this.termsLookup == null) {
+            this.termsLookup = new TermsLookup();
+        }
+        this.termsLookup.routing(lookupRouting);
         return this;
     }
 
+    /**
+     * Same as {@link #convertToBytesRefIfString} but on Iterable.
+     * @param objs the Iterable of input object
+     * @return the same input or a list of {@link BytesRef} representation if input was a list of type string
+     */
+    private static List<Object> convertToBytesRefListIfStringList(Iterable objs) {
+        if (objs == null) {
+            return null;
+        }
+        List<Object> newObjs = new ArrayList<>();
+        for (Object obj : objs) {
+            newObjs.add(convertToBytesRefIfString(obj));
+        }
+        return newObjs;
+    }
+
+    /**
+     * Same as {@link #convertToStringIfBytesRef} but on Iterable.
+     * @param objs the Iterable of input object
+     * @return the same input or a list of utf8 string if input was a list of type {@link BytesRef}
+     */
+    private static List<?> convertToStringListIfBytesRefList(Iterable<?> objs) {
```

Comment:
my bad from previous review, as I said above, change to `List<Object>` ad `Iterable<Object>`